pkg/apiserver/logsearch/pack.go

```diff
@@ -14,101 +14,57 @@
 package logsearch
 
 import (
-	"archive/tar"
 	"fmt"
-	"io"
+	"io/ioutil"
 	"net/http"
-	"os"
-	"path"
 
 	"github.com/gin-gonic/gin"
-	"github.com/pingcap/log"
-	"go.uber.org/zap"
 
 	"github.com/pingcap-incubator/tidb-dashboard/pkg/apiserver/utils"
 )
 
-func packLogsAsTarball(tasks []*TaskModel, w io.Writer) {
-	tw := tar.NewWriter(w)
-	defer tw.Close()
+func serveTaskForDownload(task *TaskModel, c *gin.Context) {
+	logPath := task.LogStorePath
+	if logPath == nil {
+		logPath = task.SlowLogStorePath
+	}
+	if logPath == nil {
+		c.Status(http.StatusBadRequest)
+		_ = c.Error(utils.ErrInvalidRequest.New("Log is not available"))
+		return
+	}
+	c.FileAttachment(*logPath, fmt.Sprintf("logs-%s.zip", task.Target.FileName()))
+}
 
+func serveMultipleTaskForDownload(tasks []*TaskModel, c *gin.Context) {
+	filePaths := make([]string, 0, len(tasks))
 	for _, task := range tasks {
-		if task.LogStorePath == nil && task.SlowLogStorePath == nil {
-			continue
-		}
-		if task.LogStorePath != nil {
-			if err := dumpLog(*task.LogStorePath, tw); err != nil {
-				log.Warn("Failed to pack log",
-					zap.Any("task", task),
-					zap.Error(err))
-				continue
-			}
+		logPath := task.LogStorePath
+		if logPath == nil {
+			logPath = task.SlowLogStorePath
 		}
-		if task.SlowLogStorePath != nil {
-			if err := dumpLog(*task.SlowLogStorePath, tw); err != nil {
-				log.Warn("Failed to pack slow log",
-					zap.Any("task", task),
-					zap.Error(err))
-				continue
-			}
+		if logPath == nil {
+			c.Status(http.StatusInternalServerError)
+			_ = c.Error(utils.ErrInvalidRequest.New("Some logs are not available"))
+			return
 		}
+		filePaths = append(filePaths, *logPath)
 	}
-}
 
-func dumpLog(savedPath string, tw *tar.Writer) error {
-	f, err := os.Open(savedPath)
+	temp, err := ioutil.TempFile("", "logs")
 	if err != nil {
-		return err
-	}
-	defer f.Close()
-	fi, err := f.Stat()
-	if err != nil {
-		return err
-	}
-	err = tw.WriteHeader(&tar.Header{
-		Name:    path.Base(savedPath),
-		Mode:    int64(fi.Mode()),
-		ModTime: fi.ModTime(),
-		Size:    fi.Size(),
-	})
-	if err != nil {
-		return err
+		c.Status(http.StatusInternalServerError)
+		_ = c.Error(err)
+		return
 	}
+	defer temp.Close()
 
-	_, err = io.Copy(tw, f)
+	err = utils.CreateZipPack(temp, filePaths)
```

Comment:
Will it be too slow to zip again when there are a lot of logs? (for example, 5 GB)

Note that previously it is a tar stream on the fly using io.Copy, which is roughly a memory copy and does not need to wait for everything to finish as well.

Reply:
Yep, this may be a problem, let me check it.

Reply:
zip provide an option to allow us not to compress the files, ref: https://stackoverflow.com/a/23717581/2998877 , has updated the code.

Reply:
stream the zip pack, ref: https://stackoverflow.com/a/57434338/2998877